git/packet_writer.go

```diff
@@ -0,0 +1,124 @@
+package git
+
+import (
+	"io"
+
+	"github.com/github/git-lfs/tools"
+)
+
+type PacketWriter struct {
+	// buf is the internal buffer of bytes used to store data given in calls
+	// to Write that don't warrant a protocol write.
+	buf []byte
+	// proto is the place where packets get written.
+	proto *protocol
+}
+
+var _ io.Writer = new(PacketWriter)
+
+// NewPacketWriter returns a new *PacketWriter, which will write to the
+// underlying data stream "w". The internal buffer is initialized with the given
+// capacity, "c".
+//
+// If "w" is already a `*PacketWriter`, it will be returned as-is.
+func NewPacketWriter(w io.Writer, c int) *PacketWriter {
+	if pw, ok := w.(*PacketWriter); ok {
+		return pw
+	}
```

Comment:
If a caller (not us, since we never make this mistake) passes tries to wrap an `io.Writer` that _is_ underneath itself a `*git.PacketWriter` to this constructor, we return that writer `w` instead of re-wrapping it and producing invalid sequences of bytes.